packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.kt

```diff
@@ -0,0 +1,1377 @@
+          promise.resolve(null)
+        }
+        fromUri.scheme == null -> {
+          // this is probably an asset embedded by the packager in resources
+          val `in` = openResourceInputStream(options["from"] as String?)
+          val out: OutputStream = FileOutputStream(uriToFile(toUri))
+          IOUtils.copy(`in`, out)
+          promise.resolve(null)
+        }
+        else -> {
+          throw IOException("Unsupported scheme for location '$fromUri'.")
+        }
+      }
+    } catch (e: Exception) {
+      e.message?.let { Log.e(TAG, it) }
+      promise.reject(e)
+    }
+  }
+
+  @Throws(IOException::class)
+  private fun transformFilesFromSAF(documentFile: DocumentFile, outputDir: File, copy: Boolean) {
+    if (!documentFile.exists()) {
+      return
+    }
+    if (!outputDir.exists() && !outputDir.mkdirs()) {
+      throw IOException("Couldn't create folder in output dir.")
+    }
+    if (documentFile.isDirectory) {
+      for (file in documentFile.listFiles()) {
+        documentFile.name?.let {
+          transformFilesFromSAF(file, File(outputDir, it), copy)
+        }
+      }
+      if (!copy) {
+        documentFile.delete()
+      }
+      return
+    }
+    documentFile.name?.let {
+      val newFile = File(outputDir.path, it)
+      context.contentResolver.openInputStream(documentFile.uri).use { `in` -> FileOutputStream(newFile).use { out -> IOUtils.copy(`in`, out) } }
+      if (!copy) {
+        documentFile.delete()
+      }
+    }
+  }
+
+  @ExpoMethod
+  fun makeDirectoryAsync(uriStr: String?, options: Map<String?, Any?>, promise: Promise) {
+    try {
+      val uri = Uri.parse(uriStr)
+      ensurePermission(uri, Permission.WRITE)
+      if (uri.scheme == "file") {
+        val file = uriToFile(uri)
+        val previouslyCreated = file.isDirectory
+        val setIntermediates = options.containsKey("intermediates") && options["intermediates"] as Boolean
+        val success = if (setIntermediates) file.mkdirs() else file.mkdir()
+        if (success || setIntermediates && previouslyCreated) {
+          promise.resolve(null)
+        } else {
+          promise.reject(
+            "ERR_FILESYSTEM_CANNOT_CREATE_DIRECTORY",
+            "Directory '$uri' could not be created or already exists."
+          )
+        }
+      } else {
+        throw IOException("Unsupported scheme for location '$uri'.")
+      }
+    } catch (e: Exception) {
+      e.message?.let { Log.e(TAG, it) }
+      promise.reject(e)
+    }
+  }
+
+  @ExpoMethod
+  fun readDirectoryAsync(uriStr: String?, options: Map<String?, Any?>?, promise: Promise) {
+    try {
+      val uri = Uri.parse(uriStr)
+      ensurePermission(uri, Permission.READ)
+      if (uri.scheme == "file") {
+        val file = uriToFile(uri)
+        val children = file.listFiles()
+        if (children != null) {
+          val result: MutableList<String> = ArrayList()
+          for (child in children) {
+            result.add(child.name)
+          }
```

Comment:
```suggestion
          val result = children.map { it.name }
```
just make sure the children is list or convert it otherwise - I'm not sure how promise serializes other iterables